### ISSUE TITLE
Using OpenCV's warpAffine

### ISSUE THREAD

**Issue by isarandi** [aleju/imgaug#79]
I found that skimage's warping function is much slower than the OpenCV implementation (by a factor of about 10 in my case). And since you discussed using other OpenCV things in another issue, I guess having OpenCV as a dependency is not a problem (perhaps it already is a dependency).

I installed skimage 0.13.1 from anaconda (and also tried the one from the Intel channel, which should be better optimized using MKL etc.) and OpenCV 3.1.0 from the menpo channel of conda.

This may be worth to check on other configurations, too, because it's quite a performance gain in my case.

**Comment by aleju** [comment]
Didn't know that OpenCV was so much faster for affine transformations. Will look into that. And yes, currently OpenCV is a dependence of the library (had hoped to get away from it, but doesn't seem like it's going to happen).

**Comment by isarandi** [comment]
I wrote a little script to measure the difference, you can try it too:

```python
import cv2
import skimage.transform
import skimage.data
import numpy as np
import timeit
import functools


transform = skimage.transform.AffineTransform(scale=(1.2, 1.2), rotation=0.3, translation=(-50,-50))
image = skimage.data.astronaut()
opencv_fn = lambda interp: cv2.warpAffine(image, transform.params[:2], image.shape[:2], flags=interp)
skimage_fn = lambda order: skimage.transform.warp(image, transform.inverse, order=order, preserve_range=True).astype(np.uint8)

for interp_cv, interp_sk in ((cv2.INTER_NEAREST, 0), (cv2.INTER_LINEAR, 1), (cv2.INTER_CUBIC, 3)):
    print('Interpolation order: ', interp_sk)
    time_cv = min(timeit.repeat(functools.partial(opencv_fn, interp_cv), repeat=10, number=500))/500
    print('  opencv:', time_cv*1000, 'ms/warp')
    time_sk = min(timeit.repeat(functools.partial(skimage_fn, interp_sk), repeat=10, number=500))/500
    print('  skimage:', time_sk*1000, 'ms/warp')
    print('  factor sk/cv:', time_sk/time_cv)
    print('------')
```

My output:

```
Interpolation order:  0
  opencv: 0.460346992011182 ms/warp
  skimage: 13.49013161601033 ms/warp
  factor sk/cv: 29.304267976367377
------
Interpolation order:  1
  opencv: 0.5996120240306482 ms/warp
  skimage: 21.203922312008217 ms/warp
  factor sk/cv: 35.362737006962384
------
Interpolation order:  3
  opencv: 2.5038309360388666 ms/warp
  skimage: 55.835901334066875 ms/warp
  factor sk/cv: 22.30018829562067
------
```
The difference is quite significant.

However, OpenCV only supports interpolation orders 0, 1 and 3. Orders 2, 4 and 5 are not available. So the skimage version would still be needed for those.

I quickly replaced the call of `tf.warp` in `iaa.AffineTransform._augment_images` by the following:

```python
if order in (0, 1, 3):
    cv_mode = dict(
        constant=cv2.BORDER_CONSTANT, edge=cv2.BORDER_REPLICATE,
        symmetric=cv2.BORDER_REFLECT_101, wrap=cv2.BORDER_WRAP,
        reflect=cv2.BORDER_REFLECT)[mode]
    cv_interp = {0: cv2.INTER_NEAREST, 1: cv2.INTER_LINEAR, 3: cv2.INTER_CUBIC}[order]

    image_warped = cv2.warpAffine(
        images[i],
        matrix[:2],
        flags=cv_interp,
        borderMode=cv_mode,
        borderValue=cval,
        dsize=images[i].shape[:2])
else:
    image_warped = tf.warp(
        images[i],
        matrix.inverse,
        order=order,
        mode=mode,
        cval=cval,
        preserve_range=True
    )
```

**Comment by isarandi** [comment]
My code should definitely be tested and extended first before merging it in. It only handles uint8 images, there are issues with the `borderValue` param, and I also should have written `dsize=(images[i].shape[1], images[i].shape[0])`. There may be some other problems with it as well. I only meant it as the general idea.

**Comment by Canxes** [comment]
I run into it too. Affine transforms take some enormous time in current implementation making real time augmentation impossible(even using multiprocessing). At least, on my machine.

**Comment by aleju** [comment]
I will look into this when I have finished my current project.
Meanwhile make sure that you use `order=[0,1]` for Affine (at least not an order of 2, 3, 4 or 5). With multiprocessing it runs fast enough for me to train models of medium size non-stop.

**Comment by Canxes** [comment]
Thanks for advice, decided to switch from order 1 to 0. About 2 times faster in my case.

**Comment by Erotemic** [comment]
FYI: Even though OpenCV's warpAffine is much faster than skimage's, it is less stable. Depending on the platform, environment, and other factors I'm not sure of, opencv functions may deadlock when used with Python's multiprocessing library. See: https://github.com/opencv/opencv/issues/5150.  

A better option may be to use Pillow's warp affine. It doesn't have all the same features that skimage / opencv have, but it is fast.

**Comment by aleju** [comment]
Thanks for the info about OpenCV's thread safety, didn't know about that yet. The library already uses some OpenCV functions in augmenters which are potentially called in background processes. Got no complains so far.
Adding pillow might be an option, but I think that requires creating a pillow.Image object and then transferring back to numpy - not sure how fast that would be.

**Comment by Erotemic** [comment]
AFAIK the opencv issue only happens for some functions. I think most cv2 functions should not cause a problem. 

Numpy arrays convert seemlessly to pillow Images, but converting back will require one extra copy operation (as far as I know, someone correct me if I'm wrong). Ultimately, I think this will be worth the speed boost you obtain over skimage (probably worth it to keep the skimage backend as an option, but also expose a new PIL backend for faster computation). 

I did a few experiments. The results show that:
* casting `np.ndarray` -> `Pillow.Image` can share memory
* casting `Pillow.Image` -> `np.ndarray` copies memory
* casting `Pillow` affine transforms are not in-place


```python

from PIL import Image
import numpy as np


# Make an zero image with a square of ones in the center
np_img = np.zeros((10, 10), dtype=np.uint8)
np_img[3:7, 3:7] = 1


"""
CLAIM:
    Conversion of an np.ndarray to a Pillow.Image shares memory.

    We will show that `pil_img` shares memory with `np_img`.
"""
# Note: pil_img.getdata() is the pil imaging core that contains the actual data
pil_img = Image.fromarray(np_img)
assert np_img.sum() == 16 and np.asarray(pil_img.getdata()).sum() == 16, (
    'np_img and pil_img should start off as the same')

"""
PROOF:
    Modifying the np_img also changes the pil_img.

    Increase np_img by 1 everywhere will increase pil_img by 1 everywhere.
"""
np_img += 1
assert np_img.sum() == 116
assert np.asarray(pil_img.getdata()).sum() == 116, (
    'the pil image data should be changed')


"""
Conversion from Pillow.Image to a np.ndarray does seem to force a copy.

Perhaps there is a different way of doing this where numpy just views the
pillow memory?
"""
new_np_img = np.asarray(pil_img.getdata())
new_np_img += 1
assert new_np_img.sum() == 216
assert np_img.sum() == 116, 'transforming the new ndarray does not change the old'


"""
However, affine transforms in pillow are not in place, so a copy would be
required either way.
"""
# note: pil uses the inverse transform
aff = np.array([[.5, 0, 2],
                [0, .5, 2]])
pil_aff_params = list(aff.ravel())
w1, h1 = np_img.shape[0:2][::-1]
imgaug = pil_img.transform((w1, h1), Image.AFFINE, pil_aff_params, resample=Image.BICUBIC)

assert np.asarray(pil_img.getdata()).sum() == 116
assert np.asarray(imgaug.getdata()).sum() == 118

# note when converting from pillow we need to re-set the shape
data = np.asarray(imgaug.getdata())
data.shape = np_img.shape
```

**Comment by Erotemic** [comment]
Here is a time comparison I did between the three warp affine methods: scikit-image, Pillow, and opencv. 
Note the y-axis is on a log scale, so a minor change in slop actually means quite a lot.  

The results show that opencv is a clear winner in terms of speed. Pillow does pretty well, note that this benchmark includes conversion to and from numpy. skimage performs the worst by a significant margin. 

However, each method does have its merits. skimage gives you access to several options that doesn't exist in opencv or pillow. Likewise opencv gives you lanczos resampling, which pil and skimage do not have. PIL is definately the most restricted in terms of its api. The only border mode is constant (I think there is a PR to let you set the fill value right now, but by default it can't do that), and it can't only will accept multichannel images in uint8 format (although you can use anything for grayscale).

I think the best approach is to have a backend option that allows the user to specify pil, cv2, or skimage when performing affine transforms. 

![figure_100](https://user-images.githubusercontent.com/3186211/34416882-a45b0dfa-ebc3-11e7-828a-28e15c4224b4.png)
![figure_101](https://user-images.githubusercontent.com/3186211/34416883-a4644316-ebc3-11e7-934f-9ab7d8c37283.png)
![figure_99](https://user-images.githubusercontent.com/3186211/34416884-a46dbca2-ebc3-11e7-889e-f5a79dff5998.png)
![figure_102](https://user-images.githubusercontent.com/3186211/34416885-a4765380-ebc3-11e7-8d91-e3ae1e83b99b.png)


For reference the code I used to perform this evaluation is here: https://github.com/Erotemic/clab/blob/839532eb58e065db027ea735b89ccec419fd38cf/clab/torch/transforms.py#L282

**Comment by aleju** [comment]
In the newest version in master, `Affine` now has a `backend` parameter, which can be set to `cv2` (OpenCV), `skimage` (scikit-image) or `auto` (chooses `cv2` wherever inputs allow it). The parameter defaults to `auto`.

OpenCV will be used in `auto` mode if the image dtype is `uint8`, there are <=4 image channels and order is among `[0, 1, 3]`. In `cv2` mode, order may also be 2 or 4, then these orders will automatically be mapped to 3 (`cv2.INTER_CUBIC`).

According to my tests, functionality seems to be essentially the same as in the scikit-image version. OpenCV supports using three values as cval though (i.e. full RGB colors instead of just grayscale). Will have to add this later as it caused problems today.

Not sure if PIL is added at any point. It doesn't seem to provide much additional value over the opencv/scikit-image combination.

**Comment by Erotemic** [comment]
Nice! I agree PIL may not be be needed at the moment, although the pytorch group seems to have something brewing called accimage, which uses a PIL-like interface, so perhaps that might be useful to use in the future. 

I see in the implementation the comment:
```
                    # opencv seems to support arrays of three cvals (ie RGB)
                    # in python2, but for some reason not in python3, so
                    # we chose one cval here
```
I believe the reason is that cval (like many params passed to opencv in python3) must be a tuple, not a list or array, for it to work both in python 2 and python 3. 

Lastly, as a comment, the work that I do mostly uses floating point image representations. Usually they  range between 0 and 1, but sometimes enforcing that bound is not possible (e.g. if one of the inputs is a height-map for satellite images). I know cv2 can handle floating point images, so it may be worth exposing that capability in the future.

**Comment by aleju** [comment]
Ah, I thought warpAffine only supported uint8. Thanks for the info. I now added float32 to the allowed dtypes and enabled RGB cvals.

**Comment by vfdev-5** [comment]
Just for info there is also a possibility to use [pillow-simd](https://github.com/uploadcare/pillow-simd) instead of PIL

**Comment by Erotemic** [comment]
Interesting. For reference I reran my benchmarks 

Without Pillow-SIMD:

![gray_float32](https://user-images.githubusercontent.com/3186211/35685083-b8aec682-0736-11e8-9cfd-eacecc356668.png)
![gray_float64](https://user-images.githubusercontent.com/3186211/35685084-b8b9a2e6-0736-11e8-8656-ba30d312a7ab.png)
![gray_uint8](https://user-images.githubusercontent.com/3186211/35685087-b92aa586-0736-11e8-9bf6-30ac86830e67.png)
![rgb_uint8](https://user-images.githubusercontent.com/3186211/35685088-b939d8d0-0736-11e8-9a92-7d1304b22660.png)

and with Pillow-SIMD

![gray_float32](https://user-images.githubusercontent.com/3186211/35685105-c7ba0a9c-0736-11e8-9ff9-6c7a882ffb3a.png)
![gray_float64](https://user-images.githubusercontent.com/3186211/35685106-c7c52846-0736-11e8-9d14-22f6846f2f68.png)
![gray_uint8](https://user-images.githubusercontent.com/3186211/35685107-c7ce9e30-0736-11e8-8da7-41838b454b38.png)
![rgb_uint8](https://user-images.githubusercontent.com/3186211/35685108-c7d88e68-0736-11e8-89f6-61fe66ad710a.png)

**Comment by vfdev-5** [comment]
👍  for the benmark, however it looks strange, seems no improvement

**Comment by Erotemic** [comment]
@vfdev-5 Code is here: 
https://gist.github.com/Erotemic/75aa64817a4f657915e2356f9c23300b

See the SpeedPlots section. Running the exact code does require my plottool library, which isn't trivial to install, however the timing code should be runnable.

**Comment by aleju** [comment]
You seem to be working with an older version of the library in your fork.
The latest version from master (and also pypi) automatically switches to OpenCV if the chosen/sampled parameters can be handled by OpenCV's affine function.
